KIPs/kip-146.md

```diff
@@ -0,0 +1,96 @@
+
+As of now, a targeted DoS is infeasible in Klaytn because it is a permissioned chain which allows only authorized validators to join the network. However, in permissionless network, a targeted DoS is a viable option for attackers. A random proposer selection can increase the difficulty of the attack by introducing an uncertainty of the upcoming proposers.
+
+Meanwhile, a few malicious validators must not be able to forge the randomness to arbitrarily determine the upcoming proposers. If not, it may increase the risk of censorship or monopolizing block rewards. Therefore, the new random proposer selection policy should be both unpredictable and verifiable.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+| Constant     | Value |
+| ------------ | ----- |
+| `FORK_BLOCK` | TBD   |
+
+### Proposer selection
+
+After `FORK_BLOCK`, the proposer selection policy `WeightedRandom` will have the new semantic.
+We define the following terms:
+
+- `proposerList`: a permutation of the validator list. `proposerList` is shuffled every `proposerRefreshInterval` blocks. This is exactly the same as Kore hardfork.
+- `seed`: `mixHash[0:8]` interpreted as the bytes of a big-endian signed 64-bit integer, where `mixHash` is the latest block's mixHash of [KIP-114](https://github.com/klaytn/kips/blob/main/KIPs/kip-114.md).
+- `picker`: a random number in the range `[0, proposerList)`. Generated by feeding `seed` to the source in [random number generation](https://pkg.go.dev/math/rand#example-package-Rand).
+
+Given `proposerList` and `picker`, we select `proposerList[picker]` as the next block proposer.
+
+This design allows a certain validator to be selected as the proposer for consecutive blocks, while guaranteeing another proposer to be selected in case of round change.
+
+See the following code that implements proposer selector in Golang:
+
+```go
+func weightedRandomProposer(valSet istanbul.ValidatorSet, round uint64, mixHash []byte) istanbul.Validator {
+	weightedCouncil, _ := valSet.(*weightedCouncil)
+	proposerList := weightedCouncil.proposers
+	numProposers := len(proposerList)
+
+    seed := int64(binary.BigEndian.Uint64(mixHash[0:8]))
+    r := rand.New(rand.NewSource(seed))
+    picker := (uint64(r.Intn(numProposers)) + round) % uint64(numProposers)
+    proposer := proposers[picker]
+
+    return proposer
+}
```

Comment:
The proposed method maintains the same mechanism for the `proposerRefreshInterval`. This means that predictions become available once the seed is determined at each interval, but they are not predictable for two consecutive intervals because the seed is not revealed yet.

To ensure unpredictability, it is suggested to update the seed value every block. The currently proposed method utilizes the mixHash value at the `proposerRefreshInterval`, which appears to nullify the previously proposed (prepared) KIPs.

Reply:
Actually, I'm going to change this logic. (That's why I changed this PR to draft)

FYI, for the previous algorithm, `mixHash` is the latest block's mixHash. Only proposerList is updated every `proposerRefreshInterval`.

Reply:
Ah, I was confused. Let me clarify. Providing additional information on how this function is called would enhance its representation.